### PR TITLE
Debug logging / CrashUtility improvements

### LINKED ISSUE #685: Improve Error-Log
@patrickgold Is there some way to automatically add device information, app version etc. to the bottom of the error log?

This would improve the quality of many future crash-issues tremendously.

### PR DESCRIPTION
**Pull request by patrickgold**
This PR is an attempt to improve the debug message logging as well as the CrashUtility.

This PR introduces a new debug log tool: Flog (= FlorisBoard Log). This utility class will gradually replace Timber and when completed, allow for a granular control of what to debug in the app. This can then also be used to enable debug logging in the beta and runtime releases, if e.g. detailed info for a specific bug report is needed.

Additionally, the CrashUtility has been improved, especially though the crash dialog which is shown after a crash. It now contains more clear instructions what to do, links to the issues instead of the new issue page (see #694) as well as generates a better formatted error log, which automatically fills out the environment info, the most important preferences and their values (such as suggestions enabled, etc.). This should reduce the countless duplicates we received in the last weeks and also make bug reports in general better formatted. See #703 for a test issue how the generated bug report looks like. "Short description" and "Steps to reproduce" are in the GitHub template, everything else is generated and automatically added to the stacktrace when using the copy button in the crash utility. Closes #685.

### CODE REVIEW

**Comment by Glitchy-Tozier** [comment]
Wow, you did a lot of coding!

I checked out the bug and it looks really nice!

One small thing i'd like to complain about: I'm not sure about removing *enviroment information* from the bug-template.

I assume you made that choice to avoid TWO *enviroment information*, but i don't like "downgrading" all other issues. 

What do you think about making two templates, a normal one and a crash one?

**Comment by Glitchy-Tozier** [comment]
(In this case the crash-log-issue would lack the *enviroment information*-heading, while the normal bug-issue would indeed contain it)

**Comment by patrickgold** [comment]
Thanks for your feedback, really appreciate it!

There are two templates, see here:

https://raw.githubusercontent.com/florisboard/florisboard/debug-crashutility-improvements/.github/ISSUE_TEMPLATE/bug_report.md

and

https://raw.githubusercontent.com/florisboard/florisboard/debug-crashutility-improvements/.github/ISSUE_TEMPLATE/bug_report_generated_error_report.md

What I could do though is renaming the "bug report generated" to crash report, which would lead to less confusion about the naming and is a bit more clear (as you directly came from the crash dialog screen). The "normal" bug report contains the environment info, while the crash report doesn't (because that's already generated by the app).

**Comment by patrickgold** [comment]
Ok I've renamed the templates and now we have two clearly marked templates:
- Bug report: https://raw.githubusercontent.com/florisboard/florisboard/debug-crashutility-improvements/.github/ISSUE_TEMPLATE/bug_report.md
- Crash report: https://raw.githubusercontent.com/florisboard/florisboard/debug-crashutility-improvements/.github/ISSUE_TEMPLATE/crash_report.md

This seems a bit more clear now. I've tried to also be consistent with the naming of the crash log as well as the template names.

**Comment by Glitchy-Tozier** [comment]
Okay, so I tried suggesting a few additional changes to the bug-template but as I'm not quite sure how to change code directly, I'll write a few comments.

**Review by Glitchy-Tozier — CHANGES_REQUESTED** [review]
``` diff
- <!--
- - Important: If you have a generated crash log to paste, use the "Crash report" template instead!
- -
- - Describe the bug in a short but concise way.
- - If you have a screenshot or screen recording of the bug, link them at the end of this issue.
- - Please search existing bug/crash reports reports to avoid creating duplicates.
- - Give your bug report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
- - Also make sure to fill out the environment information. This info is valuable when trying to fix your described bug.
- -
- - Important: should your issue be regarding an OutOfMemory crash, please don't submit this bug report and check out issue #677 (is pinned).
- -
- - Thank you for your help in making FlorisBoard better!
- -->

+ <!--
+ • Thank you for your help in making FlorisBoard better!
+ 
+ Guide to a good crash-report:
+ • Please search existing bug/crash reports reports to avoid creating duplicates.
+ • Give your bug report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
+ • Describe the bug in a short but concise way.
+ • If you have a screenshot or screen recording of the bug, link them at the end of this issue.
+ • Also make sure to fill out the environment information. This info is valuable when trying to fix your described bug.
+ -->

```


Here's a plain text-version:
```
<!--
• Thank you for your help in making FlorisBoard better!

Guide to a good crash-report:
• Please search existing bug/crash reports reports to avoid creating duplicates.
• Give your bug report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
• Describe the bug in a short but concise way.
• If you have a screenshot or screen recording of the bug, link them at the end of this issue.
• Also make sure to fill out the environment information. This info is valuable when trying to fix your described bug.
-->
```

**Comment by Glitchy-Tozier** [comment]
Crash-version:
```
- <!--
- - Important: Use this template only if you can paste a generated crash log. If you don't have such log, use the "Bug report" template instead.
- -
- - Describe what you were doing what could've led to the crash and if the crash was random or reproducible.
- - Please search existing bug/crash reports to avoid creating duplicates.
- - Give your crash report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
- -
- - Important: should your issue be regarding an OutOfMemory crash, please don't submit this bug report and check out issue #677 (is pinned).
- -
- - Thank you for your help in making FlorisBoard better!
- -->

+ <!--
+ • Thank you for your help in making FlorisBoard better!

+ Guide to a good bug-report:
+ • Please search existing bug/crash reports to avoid creating duplicates.
+ • Give your crash report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
+ • Describe what you were doing what could've led to the crash and whether the crash is random or reproducible.
+ -->
```

Plain version:
```
<!--
• Thank you for your help in making FlorisBoard better!

Guide to a good bug-report:
• Please search existing bug/crash reports to avoid creating duplicates.
• Give your crash report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
• Describe what you were doing what could've led to the crash and whether the crash is random or reproducible.
-->
```

**Comment by Glitchy-Tozier** [comment]
Sry for the wall of text. Thoughts?

**Comment by patrickgold** [comment]
Thanks for your suggesting changes to the template, they are really good. You use fewer lines to basically say the same thing and your version is a bit more understandable, so I will accept your suggestions. The only thing I see is that we probably should remove the bullet point from the "Thank you", in this structure it kinda looks out-of-place IMO.

> I understand the reason for the OOM-comment, but nobody except for like... 10 users knows what an "OutOfMemory-Crash" is supposed to be. Therefore I'd remove it to keep the comments clean. :/

This line was added by me in frustration before I started working on this PR, but I agree that most people won't really know how to decide if their issue is an OOM error.

> (Maybe) move the "Thank you" to the beginning, to make the comments more enticing to read. The only problem is that now the end is a little lackluster. And another "Thank you" feels not quite right...

Thanks to the "Guide to a good bug report" line the "thank you" on top does look pretty good and I agree that this line at the beginning would possibly lead to more people reading the few lines.

**Comment by patrickgold** [comment]
As you've seen quite a lot things to change in the bug/crash report template, could you also have a look at the crash report screen in-app (screenshot below), if the text is written understandably or if I did bloat it unnecessarily too? Thanks in advance!

crash-dialog-screen

**Comment by kj7rrv** [comment]
Could it detect OOM errors and show different instructions based on that? In Python I would write it as something like "'OutOfMemory' in stacktrace", but I'm not sure how to do it in Kotlin.On Apr 17, 2021 10:53, Patrick Goldinger ***@***.***> wrote:
As you've seen quite a lot things to change in the bug/crash report template, could you also have a look at the crash report screen in-app (screenshot below), if the text is written understandably or if I did bloat it unnecessarily too? Thanks in advance!

**Comment by patrickgold** [comment]
@kj7rrv Technically it is possible: `stacktrace.contains("OutOfMemoryError")`. I don't think though that such code should be permanent, because if now most of the OOM errors are fixed and then in the future a new OOM error appears, it should be reported again, else such issues won't get detected that easily in future releases.

**Comment by Glitchy-Tozier** [comment]
> The only thing I see is that we probably should remove the bullet point from the "Thank you", in this structure it kinda looks out-of-place IMO.

Now that i think about it, that makes a lot more sense! 

> As you've seen quite a lot things to change in the bug/crash report template, could you also have a look at the crash report screen in-app

Will look at it, give me a few hours to wake up first

File: .github/ISSUE_TEMPLATE/crash_report.md [hunk 1]
```diff
@@ -8,23 +8,17 @@ assignees: ''
...
+Guide to a good crash-report:
+• Please search existing bug/crash reports to avoid creating duplicates.
+• Give your crash report a good name (no generics like "Error" or "Crash"), so others can easily identify the topic of your issue.
+• Describe what you were doing what could've led to the crash and whether the crash is random or reproducible.
 -->
 
+
 #### Short description
-Describe the moments before the crash and what could've led to the crash.
+Describe what you were doing what could've led to the crash.
```
Comment (Glitchy-Tozier) [inline .github/ISSUE_TEMPLATE/crash_report.md:21]: I'd use "...what you were doing **that** could've..."

**Comment by Glitchy-Tozier** [comment]
> As you've seen quite a lot things to change in the bug/crash report template, could you also have a look at the crash report screen in-app (screenshot below), if the text is written understandably or if I did bloat it unnecessarily too? Thanks in advance!

It's a really good text. There's just two small changes I want to propose, and even those could be argued about.
1. After the `"If wanna report, search duplicates,"` I'd add a line break to empathize it and to separate it from the remaining text. (one is what you do BEFORE reporting, the other is HOW TO report). Only use one line-break though, no empty line.
This
```
hsnaretishernatishranti
hsnaretishrnateisrhtaeisrtharnitshratisrnhatisrhatisashri
```
, not this
```
hsnraetisnrhtaisnrthai

ahsrnetisharnetishrnatisrhatisrntaesirnthaesirnthaeitr
```

2. `"…fill out the description, steps to reproduce, and paste the…"`
Maybe that's just my English being imperfect but that list sounds weird. I'd either go for
"…fill out the description, **the** steps to reproduce, and paste the…"
or
"…fill out description, steps to reproduce, and paste the…" (leaving out the first "the")

Other than that, i can't find anything to complain about  :)

**Comment by patrickgold** [comment]
Perfect, thanks a lot for the feedback! I have incorporated it into the PR and will merge the changes into master now.

**Comment by Glitchy-Tozier** [comment]
kk  :)

**Comment by Glitchy-Tozier** [comment]
@patrickgold It looks like the change didn't go through in the bug-template.


(The crash-template exists and looks fine however.)

**Comment by patrickgold** [comment]
@Glitchy-Tozier This is a caching bug in the GitHub app then. Just tried it in the browser (on the Desktop) and the bug report looks like this:

image

**Comment by Glitchy-Tozier** [comment]
Okay, great! Ty for the fast response.

**Pull request merged**